### PR TITLE
adding gzip option

### PR DESCRIPTION
**Pull request by sunnysideup**
It is up to maintainers if you are interested in this, but I can report that it works for me with about 20% speed gains and lower server profile.

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -1,12 +1,15 @@
 {
-    "name": "silverstripe/staticpublishqueue",
+    "name": "sunnysideup/staticpublishqueue",
```
Comment (dhensby) [inline composer.json:2]: We can't change this

File: docs/examples/.htaccess.example-for-gzip [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+
+
+<IfModule mod_rewrite.c>
+	SetEnv HTTP_MOD_REWRITE On
+	RewriteEngine On
+
+	## CONFIG FOR STATIC PUBLISHING
+	# Cached content - sub-pages (site in the root of a domain)
+    RewriteCond "%{HTTP:Accept-encoding}" "gzip"
```
Comment (dhensby) [inline docs/examples/.htaccess.example-for-gzip:37]:
```suggestion
	RewriteCond "%{HTTP:Accept-encoding}" "gzip"
```

File: docs/examples/.htaccess.example-for-gzip [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	RewriteCond %{REQUEST_METHOD} ^GET|HEAD$
+	RewriteCond %{QUERY_STRING} ^$
+	RewriteCond %{REQUEST_URI} /(.*[^/])/?$
+	RewriteCond %{DOCUMENT_ROOT}/cache/%1.html.gz -f
+	RewriteCond %{REQUEST_FILENAME} !-f
+	RewriteRule .* /cache/%1.html.gz [L]
+
+	# Cached content - homepage (site in root of a domain)
+    RewriteCond "%{HTTP:Accept-encoding}" "gzip"
```
Comment (dhensby) [inline docs/examples/.htaccess.example-for-gzip:46]:
```suggestion
	RewriteCond "%{HTTP:Accept-encoding}" "gzip"
```

File: docs/examples/.htaccess.example-for-gzip [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	RewriteCond %{QUERY_STRING} ^$
+	RewriteCond %{REQUEST_URI} ^/?$
+	RewriteCond %{DOCUMENT_ROOT}/cache/index.html.gz -f
+	RewriteCond %{REQUEST_FILENAME} !-f
+	RewriteRule .* /cache/index.html.gz [L]
+
+	# Config for site in a subdirectory
+	# (START remove the following four rules if site is on root of a domain. E.g. test.com rather than test.com/my-site)
+	# Cached content - sub-pages (site in sub-directory)
+    RewriteCond "%{HTTP:Accept-encoding}" "gzip"
```
Comment (dhensby) [inline docs/examples/.htaccess.example-for-gzip:57]:
```suggestion
	RewriteCond "%{HTTP:Accept-encoding}" "gzip"
```

File: docs/examples/.htaccess.example-for-gzip [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	RewriteCond %{REQUEST_METHOD} ^GET|HEAD$
+	RewriteCond %{QUERY_STRING} ^$
+	RewriteCond %{REQUEST_URI} /(.*[^/])/(.*[^/])/?$
+	RewriteCond %{DOCUMENT_ROOT}/%1/cache/%2.html.gz -f
+	RewriteCond %{REQUEST_FILENAME} !-f
+	RewriteRule .* cache/%2.html.gz [L]
+
+	# Cached content - homepage (site in sub-directory)
+    RewriteCond "%{HTTP:Accept-encoding}" "gzip"
```
Comment (dhensby) [inline docs/examples/.htaccess.example-for-gzip:66]:
```suggestion
	RewriteCond "%{HTTP:Accept-encoding}" "gzip"
```

File: includes/functions.php [hunk 6]
```diff
@@ -2,22 +2,24 @@
 
 namespace SilverStripe\StaticPublishQueue;
 
-if (!function_exists('SilverStripe\\StaticPublishQueue\\URLtoPath')) {
+if (! function_exists('SilverStripe\\StaticPublishQueue\\URLtoPath')) {
```
Comment (dhensby) [inline includes/functions.php:5]:
```suggestion
if (!function_exists('SilverStripe\\StaticPublishQueue\\URLtoPath')) {
```

File: includes/functions.php [hunk 7]
```diff
@@ -2,22 +2,24 @@
...
+        $url = str_replace('?stage=Live', '?', $url);
+
         // parse_url() is not multibyte safe, see https://bugs.php.net/bug.php?id=52923.
         // We assume that the URL hsa been correctly encoded either on storage (for SiteTree->URLSegment),
         // or through URL collection (for controller method names etc.).
         $urlParts = @parse_url($url);
 
         // query strings are not yet supported so we need to bail is there is one present
-        if (!empty($urlParts['query'])) {
+        if (! empty($urlParts['query'])) {
```
Comment (dhensby) [inline includes/functions.php:16]:
```suggestion
        if (!empty($urlParts['query'])) {
```

File: includes/functions.php [hunk 8]
```diff
@@ -26,10 +28,10 @@ function URLtoPath($url, $baseURL = '', $domainBasedCaching = false)
         // Normalize URLs
         $urlSegment = trim($urlSegment, '/');
 
-        $filename = $urlSegment ?: "index";
+        $filename = $urlSegment ?: 'index';
 
         if ($domainBasedCaching) {
-            if (!$urlParts) {
+            if (! $urlParts) {
```
Comment (dhensby) [inline includes/functions.php:34]:
```suggestion
            if (!$urlParts) {
```

File: includes/functions.php [hunk 9]
```diff
@@ -38,14 +40,14 @@ function URLtoPath($url, $baseURL = '', $domainBasedCaching = false)
...
-        if ($dirName != '/' && $dirName != '.') {
+        if ($dirName !== '/' && $dirName !== '.') {
             $prefix = $dirName . '/';
         }
         return $prefix . basename($filename);
     }
 }
 
-if (!function_exists('SilverStripe\\StaticPublishQueue\\PathToURL')) {
+if (! function_exists('SilverStripe\\StaticPublishQueue\\PathToURL')) {
```
Comment (dhensby) [inline includes/functions.php:50]:
```suggestion
if (!function_exists('SilverStripe\\StaticPublishQueue\\PathToURL')) {
```

File: includes/staticrequesthandler.php [hunk 10]
```diff
@@ -2,8 +2,7 @@
 
 namespace SilverStripe\StaticPublishQueue;
 
-return function($cacheDir, $urlMapping = null)
-{
+return function ($cacheDir, $urlMapping = null) {
```
Comment (dhensby) [inline includes/staticrequesthandler.php:5]:
```suggestion
return function ($cacheDir, $urlMapping = null)
{
```

File: includes/staticrequesthandler.php [hunk 11]
```diff
@@ -18,19 +17,26 @@
...
 
-
     if (is_callable($urlMapping)) {
         $path = $urlMapping($url);
     } else {
         $path = URLtoPath($url);
     }
 
-    if (!$path) {
+    if (! $path) {
```
Comment (dhensby) [inline includes/staticrequesthandler.php:26]:
```suggestion
    if (!$path) {
```

File: includes/staticrequesthandler.php [hunk 12]
```diff
@@ -18,19 +17,26 @@
...
     }
 
     $cachePath = $cacheDir . DIRECTORY_SEPARATOR . $path;
-
+    $hasHTMLFile = false;
+    $hasGZIPFile = false;
+    if (file_exists($cachePath . '.html')) {
+        $hasHTMLFile = true;
+        $cachePath .= '.html';
+    } elseif (file_exists($cachePath . '.html.gz')) {
```
Comment (dhensby) [inline includes/staticrequesthandler.php:36]: Should it not check for the gzip first and serve that as a priority? Could both exist at the same time?

File: includes/staticrequesthandler.php [hunk 13]
```diff
@@ -44,26 +50,26 @@
 
     if (file_exists($cachePath . '.php')) {
         $cacheConfig = require $cachePath . '.php';
-    } elseif (!file_exists($cachePath . '.html')) {
+    } elseif (! $hasGZIPFile && ! $hasHTMLFile) {
```
Comment (dhensby) [inline includes/staticrequesthandler.php:53]:
```suggestion
    } elseif (!$hasGZIPFile && !$hasHTMLFile) {
```

File: includes/staticrequesthandler.php [hunk 14]
```diff
@@ -44,26 +50,26 @@
...
         return false;
     }
     header('X-Cache-Hit: ' . date(\DateTime::COOKIE));
-    if (!empty($cacheConfig['responseCode'])) {
+    if (! empty($cacheConfig['responseCode'])) {
```
Comment (dhensby) [inline includes/staticrequesthandler.php:57]:
```suggestion
    if (!empty($cacheConfig['responseCode'])) {
```

File: includes/staticrequesthandler.php [hunk 15]
```diff
@@ -44,26 +50,26 @@
...
         header('HTTP/1.1 ' . $cacheConfig['responseCode']);
     }
-    if (!empty($cacheConfig['headers'])) {
+    if (! empty($cacheConfig['headers'])) {
```
Comment (dhensby) [inline includes/staticrequesthandler.php:60]:
```suggestion
    if (!empty($cacheConfig['headers'])) {
```

File: includes/staticrequesthandler.php [hunk 16]
```diff
@@ -44,26 +50,26 @@
...
         foreach ($cacheConfig['headers'] as $header) {
             header($header, true);
         }
     }
-    if (file_exists($cachePath . '.html')) {
-        $etag = '"' . md5_file($cachePath . '.html') . '"';
-        if (!empty($_SERVER['HTTP_IF_NONE_MATCH']) && $_SERVER['HTTP_IF_NONE_MATCH'] == $etag) {
+    if ($hasHTMLFile || $hasGZIPFile) {
+        $etag = '"' . md5_file($cachePath) . '"';
+        if (! empty($_SERVER['HTTP_IF_NONE_MATCH']) && $_SERVER['HTTP_IF_NONE_MATCH'] === $etag) {
```
Comment (dhensby) [inline includes/staticrequesthandler.php:67]:
```suggestion
        if (!empty($_SERVER['HTTP_IF_NONE_MATCH']) && $_SERVER['HTTP_IF_NONE_MATCH'] === $etag) {
```

File: src/Contract/StaticPublisher.php [hunk 17]
```diff
@@ -8,16 +8,22 @@ interface StaticPublisher
      * @param string $url
      * @return array A result array
      */
-    public function publishURL($url, $forcePublish = false);
+    public function publishURL(string $url, ?bool $forcePublish = false): array;
```
Comment (dhensby) [inline src/Contract/StaticPublisher.php:11]: I think this is a breaking change and we should try to avoid that

File: src/Contract/StaticPublishingTrigger.php [hunk 18]
```diff
@@ -9,22 +9,21 @@
...
 interface StaticPublishingTrigger
 {
-
     /**
      * Provides an SS_List of StaticallyPublishable objects which need to be regenerated.
      *
      * @param array $context An associative array with extra engine-specific information.
      *
-     * @return SS_List
+     * @return array|SS_List
```
Comment (dhensby) [inline src/Contract/StaticPublishingTrigger.php:17]: You're forcing an array type in the argument declaration but this PHPDoc claims to accept SS_List as well. I think we need to remove the type declarations...

File: src/Extension/Engine/SiteTreePublishingEngine.php [hunk 19]
```diff
@@ -63,7 +63,7 @@ public function setToUpdate($toUpdate)
     }
 
     /**
-     * @param $toDelete
+     * @param mixed $toDelete
```
Comment (dhensby) [inline src/Extension/Engine/SiteTreePublishingEngine.php:66]:
```suggestion
     * @param array $toDelete
```

File: src/Extension/Engine/SiteTreePublishingEngine.php [hunk 20]
```diff
@@ -138,7 +138,7 @@ public function collectChanges($context)
     public function flushChanges()
     {
         $queue = QueuedJobService::singleton();
-        if (!empty($this->toUpdate)) {
+        if (! empty($this->toUpdate)) {
```
Comment (dhensby) [inline src/Extension/Engine/SiteTreePublishingEngine.php:141]:
```suggestion
        if (!empty($this->toUpdate)) {
```

File: src/Extension/Engine/SiteTreePublishingEngine.php [hunk 21]
```diff
@@ -148,15 +148,15 @@ public function flushChanges()
...
                 ]);
 
                 $queue->queueJob($job);
             }
-            $this->toUpdate = array();
+            $this->toUpdate = [];
         }
 
-        if (!empty($this->toDelete)) {
+        if (! empty($this->toDelete)) {
```
Comment (dhensby) [inline src/Extension/Engine/SiteTreePublishingEngine.php:159]:
```suggestion
        if (!empty($this->toDelete)) {
```

File: src/Job/DeleteStaticCacheJob.php [hunk 22]
```diff
@@ -28,12 +28,12 @@ public function process()
...
         $chunkSize = self::config()->get('chunk_size');
         $count = 0;
-        foreach ($this->jobData->URLsToProcess as $url => $priority) {
+        foreach (array_keys($this->jobData->URLsToProcess) as $url) {
             if (++$count > $chunkSize) {
                 break;
             }
             $meta = Publisher::singleton()->purgeURL($url);
-            if (!empty($meta['success'])) {
+            if (! empty($meta['success'])) {
```
Comment (dhensby) [inline src/Job/DeleteStaticCacheJob.php:36]:
```suggestion
            if (!empty($meta['success'])) {
```

File: src/Job/GenerateStaticCacheJob.php [hunk 23]
```diff
@@ -22,12 +22,12 @@ public function process()
...
         $chunkSize = self::config()->get('chunk_size');
         $count = 0;
-        foreach ($this->jobData->URLsToProcess as $url => $priority) {
+        foreach (array_keys($this->jobData->URLsToProcess) as $url) {
             if (++$count > $chunkSize) {
                 break;
             }
             $meta = Publisher::singleton()->publishURL($url, true);
-            if (!empty($meta['success'])) {
+            if (! empty($meta['success'])) {
```
Comment (dhensby) [inline src/Job/GenerateStaticCacheJob.php:30]:
```suggestion
            if (!empty($meta['success'])) {
```

File: src/Publisher.php [hunk 24]
```diff
@@ -59,7 +58,7 @@ public function generatePageResponse($url)
             $url = Director::absoluteURL($url);
         }
         $urlParts = parse_url($url);
-        if (!empty($urlParts['query'])) {
+        if (! empty($urlParts['query'])) {
```
Comment (dhensby) [inline src/Publisher.php:61]:
```suggestion
        if (!empty($urlParts['query'])) {
```

File: src/Publisher.php [hunk 25]
```diff
@@ -145,7 +144,7 @@ protected function generatePHPCacheFile($response)
         ];
 
         foreach ($response->getHeaders() as $header => $value) {
-            if (!in_array($header, ['cache-control'])) {
+            if (! in_array($header, ['cache-control'], true)) {
```
Comment (dhensby) [inline src/Publisher.php:147]:
```suggestion
            if (!in_array($header, ['cache-control'], true)) {
```

File: src/Publisher/FilesystemPublisher.php [hunk 26]
```diff
@@ -44,7 +48,7 @@ public function getDestFolder()
     public function setFileExtension($fileExtension)
     {
         $fileExtension = strtolower($fileExtension);
-        if (!in_array($fileExtension, ['html', 'php'])) {
+        if (! in_array($fileExtension, ['html', 'php'], true)) {
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:51]:
```suggestion
        if (!in_array($fileExtension, ['html', 'php'], true)) {
```

File: src/Publisher/FilesystemPublisher.php [hunk 27]
```diff
@@ -19,7 +18,12 @@ class FilesystemPublisher extends Publisher
     /**
      * @var string
      */
-    protected $fileExtension = 'php';
+    protected $fileExtension = 'html';
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:21]: What implication does this change have?

File: src/Publisher/FilesystemPublisher.php [hunk 28]
```diff
@@ -63,11 +67,11 @@ public function getFileExtension()
         return $this->fileExtension;
     }
 
-    public function purgeURL($url)
+    public function purgeURL(string $url): array
     {
-        if (!$url) {
-            user_error("Bad url:" . var_export($url, true), E_USER_WARNING);
-            return;
+        if (! $url) {
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:72]:
```suggestion
        if (!$url) {
```

File: src/Publisher/FilesystemPublisher.php [hunk 29]
```diff
@@ -84,20 +88,27 @@ public function purgeURL($url)
...
      * @param string $url
      * @return array A result array
      */
-    public function publishURL($url, $forcePublish = false)
+    public function publishURL(string $url, ?bool $forcePublish = false): array
     {
-        if (!$url) {
-            user_error("Bad url:" . var_export($url, true), E_USER_WARNING);
-            return;
+        if (! $url) {
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:104]:
```suggestion
        if (!$url) {
```

File: src/Publisher/FilesystemPublisher.php [hunk 30]
```diff
@@ -145,6 +181,10 @@ protected function publishPage($response, $url)
     {
         $success = true;
         if ($path = $this->URLtoPath($url)) {
+            // little hack to make sure we do not include pages with live forms.
+            if (stripos($response->getBody(), 'name="SecurityID"')) {
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:185]: This feels a bit out of scope and is a very fragile implementation because `SecurityID` is configurable. I'd prefer this to be removed from this PR

File: src/Publisher/FilesystemPublisher.php [hunk 31]
```diff
@@ -174,8 +215,25 @@ protected function saveToPath($content, $filePath)
...
+
+        return file_exists($publishPath);
+    }
+
+    protected function compressFile(string $publishPath): string
+    {
+        //we keep the html file for now ... so use second parameter to achieve this.
+        $publishPathGZipped = $publishPath . '.gz';
+        exec('rm ' . $publishPathGZipped . ' -rf');
+        exec('gzip ' . $publishPath);
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:233]: What if gzip is unavailable?

Why can't we gzip with PHP rather than using `exec`? This use of exec makes me a bit worried.

File: src/Publisher/FilesystemPublisher.php [hunk 32]
```diff
@@ -174,8 +215,25 @@ protected function saveToPath($content, $filePath)
...
+        }
+
+        return file_exists($publishPath);
+    }
+
+    protected function compressFile(string $publishPath): string
+    {
+        //we keep the html file for now ... so use second parameter to achieve this.
+        $publishPathGZipped = $publishPath . '.gz';
+        exec('rm ' . $publishPathGZipped . ' -rf');
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:232]: This should *never* be a directory so the use of `-r` should not be necessary and exposes this call to a larger amount of power than is required

File: src/Publisher/FilesystemPublisher.php [hunk 33]
```diff
@@ -174,8 +215,25 @@ protected function saveToPath($content, $filePath)
...
+        exec('gzip ' . $publishPath);
+        exec('chmod 0777 ' . $publishPathGZipped);
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:234]: Why do we want all users to be able to write to this file?

**Review by dhensby — CHANGES_REQUESTED** [review]
There are a fair few linting changes needed. The tests are failing, perhaps because the default publish format has changed from php -> html. The PHP files should be able to deliver the gzipped data just as apache can, which would allow dynamic headers *and* a performance increase - that feature appears to be missing from this PR?

File: composer.json (hunk 1 above)
Comment (sunnysideup) [reply composer.json:2]: ;-) Sorry, I wanted to use this new version so I changed it temporarily so that I could publish on packagist.

File: src/Publisher/FilesystemPublisher.php (hunk 32 above)
Comment (sunnysideup) [reply src/Publisher/FilesystemPublisher.php:232]: Absolutely - DONE

File: src/Publisher/FilesystemPublisher.php (hunk 33 above)
Comment (sunnysideup) [reply src/Publisher/FilesystemPublisher.php:234]: Absolutely - DONE

File: src/Publisher/FilesystemPublisher.php (hunk 27 above)
Comment (sunnysideup) [reply src/Publisher/FilesystemPublisher.php:21]: reversed!

File: src/Publisher/FilesystemPublisher.php (hunk 30 above)
Comment (sunnysideup) [reply src/Publisher/FilesystemPublisher.php:185]: made it optional - I still think this is a very easy way to avoid pages with forms.

**Review by sunnysideup — COMMENTED** [review]
DONE

**Comment by sunnysideup** [comment]
Ok - so I have done another big cleanup of all the points you raised - awesome.  This is a very optional PR, but if you are keen then I am keen to make it happen!

File: includes/staticrequesthandler.php [hunk 34]
```diff
@@ -18,19 +17,25 @@
...
 
-    if (!$path) {
+    if (! $path) {
         return false;
     }
 
     $cachePath = $cacheDir . DIRECTORY_SEPARATOR . $path;
-
+    $hasCachedFile = true;
+    if (file_exists($cachePath . '.html.gz')) {
```
Comment (dhensby) [inline includes/staticrequesthandler.php:32]: I think this needs to be a bit more robust. We shouldn't look for the gzipped version if there is no valid `Accept-Encoding` header. We should also always check the html file exists and calculate the ETag from that.

File: src/Contract/StaticPublishingTrigger.php (hunk 18 above)
Comment (dhensby) [reply src/Contract/StaticPublishingTrigger.php:17]: Sorry, bit of confusion - this is the `@returns` so I think it was fine as it was.

File: src/Publisher/FilesystemPublisher.php [hunk 35]
```diff
@@ -145,16 +166,25 @@ protected function publishPage($response, $url)
     {
         $success = true;
         if ($path = $this->URLtoPath($url)) {
+            $body = $response->getBody();
+            if ($this->Config()->get('lazy_form_recognition')) {
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:170]: wrong case
```suggestion
            if ($this->config()->get('lazy_form_recognition')) {
```

File: src/Publisher/FilesystemPublisher.php [hunk 36]
```diff
@@ -174,8 +204,35 @@ protected function saveToPath($content, $filePath)
...
         Filesystem::makeFolder(dirname($publishPath));
+        $successWithPublish = rename($temporaryPath, $publishPath);
+        if ($successWithPublish) {
+            if (FilesystemPublisher::config()->get('use_gzip_compression')) {
+                $publishPath = $this->compressFile($publishPath);
+            }
+        }
 
-        return rename($temporaryPath, $publishPath);
+        return file_exists($publishPath);
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:214]: I feel like if the gzip compression fails we shouldn't return false - I think it should be an optional step.

File: src/Publisher/FilesystemPublisher.php [hunk 37]
```diff
@@ -174,8 +204,35 @@ protected function saveToPath($content, $filePath)
...
+        $publishPathGZipped = $publishPath . '.gz';
+        // remove
+        unlink($publishPathGZipped);
+        // open a new one
+        $fp = fopen($publishPathGZipped, 'w');
+        // write it
+        fwrite($fp, $gzdata);
+        // close it
+        fclose($fp);
+        @chmod($file, 0664);
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:233]: `$file` is undefined

File: src/Publisher/FilesystemPublisher.php [hunk 38]
```diff
@@ -174,8 +204,35 @@ protected function saveToPath($content, $filePath)
...
+        }
 
-        return rename($temporaryPath, $publishPath);
+        return file_exists($publishPath);
+    }
+
+    protected function compressFile(string $publishPath): string
+    {
+        // read original
+        $data = implode('', file($publishPath));
```
Comment (dhensby) [inline src/Publisher/FilesystemPublisher.php:220]: This removes the new lines, doesn't it? perhaps `file_get_contents` would be better?

**Review by dhensby — CHANGES_REQUESTED** [review]
OK - I've had a good look at this now, rebasing and squashing things down from 53 commits to 25, there's still some tidy up to do but this has given me a better understanding of what is going on.

The `use_gzip_compression` needs to be part of `Publisher` and not `FileSystemPublisher`, it should not be a feature solely of the FileSystemPublisher, any implementor of `Publisher` should be able to implement it (optionally).

**Comment by dhensby** [comment]
I'm pulling out the linting / bugs into sepearate PRs so they can be evaluated on their own merit and so that this PR is focused on the GZipping aspects only

**Comment by dhensby** [comment]
OK - latest update.

I've split out all the changes into separate PRs, that allows us to concentrate feedback / work in the right places and to stay focused on getting each feature merged.

I've got the linting changes merged already, which is great news.

So the main feedback for the gzipping is:

1. [ ] We need to make sure the appropriate Accept-Encoding header has been received by the server.
2. [ ] The ETag should be calculated from the HTML file always (if the browser sends a request not accepting gzip and we give them the file uncompressed and then they ask again with the etag and saying they can accept gzip they should still get a 304)
3. [ ] The response needs to include a content-encoding header
4. [ ] We need some tests

**Comment by dhensby** [comment]
Right, this is quite old now and lots of the changes from the (original) PR have been picked out and merged. There's a lot of conflicts to get over which I'm not entirely enthusiastic about resolving... Plus I'm still sceptical about this server-side gzipping, though I'd accept taking it in if there were gzipping happening via a PHP extension and not via a CLI execution..

So in light of this I'm going to close this PR
